Remove some allocations from X509 library

stephentoub opened the pull request:
This addresses three of the items listed in https://github.com/dotnet/corefx/issues/1995, mostly related to reducing allocations:

1. "Internal.Cryptography.Helpers::DecodeHexString should skip whitespace, not use a string-allocating interior trim method"
2. "X509Certificate.ToString() and X509Certificate2.ToString() do a lot of allocations, try to reduce them"
3. "SyncRoot allocates a new object every call (which is thrown away on all calls aside from the first)"

The third one also addresses a bug: SyncRoot was returning null the first time it was accessed.

stephentoub commented:
cc: @bartonjs, @KrzysztofCwalina

KrzysztofCwalina commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs:
```diff
@@ -83,6 +72,14 @@ private static byte HexToByte(char val)
                 return 0xFF;
         }
 
+        private static char ByteToHex(byte b)
```
Should it be called NibbleToHex? And why does it return "?" in what seems like an error case?

KrzysztofCwalina commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Certificate.cs:
```diff
@@ -262,32 +262,42 @@ public virtual String ToString(bool fVerbose)
             byte[] serialNumber = this.GetSerialNumber();
             Array.Reverse(serialNumber);
-            sb.Append(serialNumber.ToHexStringUpper());
+            sb.AppendLine(serialNumber.ToHexStringUpper());
```
It seems to me like ToHexStringUpper should return char[] instead of string and then the char[] should be appended here.

stephentoub replied:
Good idea.

bartonjs commented:
The pedant in me cares about Byte vs Nybble.  The pragmatist in me doesn't care.
Pragmatist: "Hey, pedant, look over there!" ::points::
(Pedant turns to see the the intended target)
Pragmatist: LGTM

KrzysztofCwalina commented:
@bartonjs, yeah, normally I would not care, but there seems to be an error condition there. If you pass more than a nibble, you will get "?" back. I though the naming might make it clearer.

bartonjs commented:
@KrzysztofCwalina Oh, I agree :smile:, which is why I had also pointed it out.  I just won't withhold my blessing based on the naming of those two private methods.

stephentoub commented:
Thanks, guys.  While I was just trying to mirror the function that already existed, both in name and in function, since you both had the same objection, I'll go ahead and make the change.  I'm hesitant to change the implementation of the existing function in case some code somewhere depended on it, but I'll take a look to see if I can easily prove it doesn't matter.

KrzysztofCwalina commented:
BTW, I don't feel super strongly about the name; treat it as a suggestion. I do care a bit more about converting char[] to string so that it can immediately be converted back to chars inserted into the string builder.

stephentoub commented:
Yeah, the char[] to string was just an oversight on my part: I was focused on re-implementing the body while maintaining the signature (which already returned string), and didn't go look at the call sites to see how it was used.

stephentoub commented:
Thanks for the review, guys.  I fixed up the name and the implementation of NibbleToHex, avoided roundtripping through a string for the two call sites where a char[] was acceptable (there were six others that needed a string), and per your request Jeremy, added a commit to remove the extraneous "this." usage.

ellismg commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs:
```diff
@@ -16,59 +14,54 @@ public static byte[] CloneByteArray(this byte[] src)
             {
-                hex[index] = (byte)((HexToByte(hexString[i]) << 4) | HexToByte(hexString[i + 1]));
-                i += 2;
+                if (Char.IsWhiteSpace(s[i]))
```
NIT: char not Char, correct?

stephentoub replied:
Ugh, yeah, I was just copying and pasting what was there.  I'll fix it.

KrzysztofCwalina commented:
Looks good!
